README.md
# gowit

[![Build Status](https://travis-ci.org/dragonraid/gowit.svg?branch=master)](https://travis-ci.org/dragonraid/gowit)

`gowit` is the Go SDK for [Wit.ai](https://wit.ai/)

**This is WIP**

## Install
Using `go get`:
```
go get github.com/dragonraid/gowit
```	
From source:
```
git clone https://github.com/dragonraid/gowit $GOPATH/src/github.com/dragonraid/
go install $GOPATH/src/github.com/dragonraid/gowit
```	
## Usage:

For minimal working example you need to specify `WIT_API_TOKEN` environment variable. Then you can do:
```
witAPI := wit.New()
body, _, err := witApi.Message("YOUR MESSAGE").Do()
```

See the also `examples` directory

## API

### Versioning
The default API version is `20180527`. You can target a specific version by setting the env variable `WIT_API_VERSION`. 

### Overview

`gowit` provides interaction with following endpoints:

* `Message` - [the Wit message API](https://wit.ai/docs/http/20170307#get--message-link)


## Optional settings
You have two options how to setup your client behaviour
* Environment variables
	* `WIT_API_URL`     (defaults to `https://api.wit.ai`)
	* `WIT_API_VERSION` (defaults to `20180527`)
	* `WIT_API_VERBOSE` (defaults to `false`) 
	
	
* Wit struct 
```
witApi := &wit.Wit{
	Token:   "<WIT_API_TOKEN>",
	URL:     "<WIT_API_URL>",
	Version: "<WIT_API_VERSION>",
	Verbose: "<WIT_API_VERBOSE>",
}
```


CHANGELOG.md
# Changelog

## 0.2.0

### Added
* `wit/message/SetParameters`: Enables to populate request url with optional parameters.


wit/wit_test.go
package wit_test

import (
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"
	"time"

	"github.com/dragonraid/gowit/wit"
)

var config = []struct {
	Input       *wit.Wit
	Expected    error
	Description string
}{
	{
		Input: &wit.Wit{
			Token:   "WitAPItoken",
			URL:     "https://api.wit.ai",
			Version: "20180711",
			Verbose: "false",
		},
		Expected:    nil,
		Description: "Correct configuration",
	},
	{
		Input: &wit.Wit{
			Token:   "",
			URL:     "https://api.wit.ai",
			Version: "20180711",
			Verbose: "false",
		},
		Expected:    wit.ErrInvalidToken,
		Description: "Invalid API token",
	},
	{
		Input: &wit.Wit{
			Token:   "WitAPItoken",
			URL:     "https://api.wit.ai",
			Version: "2018071111",
			Verbose: "false",
		},
		Expected: &time.ParseError{
			Layout:     "20060102",
			Value:      "2018071111",
			LayoutElem: "",
			ValueElem:  "11",
			Message:    ": extra text: " + "11"},
		Description: "Invalid version format",
	},
	{
		Input: &wit.Wit{
			Token:   "WitAPItoken",
			URL:     "https://api.wit.ai",
			Version: "30180111",
			Verbose: "false",
		},
		Expected:    &wit.ErrInvalidVersion{Version: "30180111"},
		Description: "Invalid version",
	},
	{
		Input: &wit.Wit{
			Token:   "WitAPItoken",
			URL:     "https://api.wit.ai",
			Version: "20180711",
			Verbose: "true",
		},
		Expected:    nil,
		Description: "Verbose true",
	},
	{
		Input: &wit.Wit{
			Token:   "WitAPItoken",
			URL:     "https://api.wit.ai",
			Version: "20180711",
			Verbose: "",
		},
		Expected:    wit.ErrInvalidVerbose,
		Description: "Invalid Verbose parameter",
	},
}

// RequestHandler type
type RequestHandler func(w http.ResponseWriter, r *http.Request)

// TestWit f
func TestWit(t *testing.T) {
	for _, conf := range config {
		t.Run(conf.Description, func(t *testing.T) {
			if err := conf.Input.SanityCheck(); reflect.TypeOf(err) == reflect.TypeOf(conf.Expected) {
				if err == nil {
				} else if err.Error() != conf.Expected.Error() {
					t.Errorf("Expected \"%s\", but got \"%s\"\n",
						conf.Expected.Error(), err.Error())
				}
			} else {
				t.Errorf("Expected \"%T\", but got \"%T\"\n",
					conf.Expected, err)
			}
		})
	}
}

// WitTestRequest f
func WitTestRequest(method, params string, fn RequestHandler) (*httptest.ResponseRecorder, error) {
	_, request, err := config[0].Input.CreateRequest(method, params)
	responseRecorder := httptest.NewRecorder()
	handler := http.HandlerFunc(fn)
	handler.ServeHTTP(responseRecorder, request)
	return responseRecorder, err
}


wit/message_test.go
package wit_test

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"testing"

	"github.com/dragonraid/gowit/wit"
)

// MessageTestStruct struct
type MessageTestStruct struct {
	Input       *wit.Message
	Response    string
	Expected    []error
	Description string
}

var message = []MessageTestStruct{
	{
		Input: config[0].Input.Message("Message test response"),
		Response: `{
                              "_text": "Message test response", 
                              "entities": {
                                "datetime": [
                                  {
                                    "confidence": 0.9,
                                    "value": "test"
                                  }
                                ],
                                "intent": "date"
                              },
                              "msg_id": "1a2b3c"
                           }`,
		Expected:    []error{nil, nil},
		Description: "Correct response",
	},
	{
		Input: config[0].Input.Message("Message test response"),
		Response: `{
                              "_text": "Message test response", 
                              "entities": {
                                  9:false,
                              },
                              "msg_id": "1a2b3c"
                            }`,
		Expected:    []error{nil, fmt.Errorf("invalid character '9' looking for beginning of object key string")},
		Description: "Incorrect response",
	},
}

// MessageHandler function mocks wit.ai/messages endpoint
func (m *MessageTestStruct) MessageHandler(w http.ResponseWriter, r *http.Request) {
	_, _ = io.WriteString(w, m.Response)
	//w.Write(payload)
}

// TestMessageHandler function uses MessageHandler to perform specific usecase tests
func TestMessageHandler(t *testing.T) {
	for _, msg := range message {
		t.Run(msg.Description, func(t *testing.T) {
			responseRecorder, err := WitTestRequest("GET", msg.Input.Params, msg.MessageHandler)
			if err != msg.Expected[0] {
				t.Errorf("An error occured: \"%s\"", err)
			}
			var responseBody wit.MessageResponse
			err = json.Unmarshal(responseRecorder.Body.Bytes(), &responseBody)
			if err == nil {
			} else if err.Error() != msg.Expected[1].Error() {
				t.Errorf("An error occured: \"%s\"", err)
			}
		})
	}

}


wit/message.go
package wit

import (
	"encoding/json"
	"net/url"
)

// Message struct
type Message struct {
	WitConf *Wit
	Params  string
	// params[]string
}

// MessageResponse struct to map response body of wit.ai/message endpoint
type MessageResponse struct {
	Text     string                 `json:"_text"`
	Entities map[string]interface{} `json:"entities"`
	MsgID    string                 `json:"msg_id"`
}

// Do function makes actual call to wit.ai/message endpoint and
// retrieves its response
func (m *Message) Do() (MessageResponse, int, error) {
	var responseBody MessageResponse
	body, statusCode, err := m.WitConf.MakeRequest("GET", m.Params)
	if err != nil {
		return responseBody, 0, err
	}
	err = json.Unmarshal(body, &responseBody)
	if err != nil {
		return responseBody, 0, err
	}
	return responseBody, statusCode, nil
}

// SetParameters function sets additional parameters to url message
func (m *Message) SetParameters(parameters string) {
	encodedParameters := &url.URL{Path: parameters}
	m.Params += "&" + encodedParameters.String()
}


wit/wit.go
package wit

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"time"
)

// Exported variables
var (
	ErrInvalidToken   = fmt.Errorf("Please use valid token")
	ErrInvalidVerbose = fmt.Errorf("WIT_API_VERBOSE must be set to true or false")
)

// ErrInvalidVersion error
type ErrInvalidVersion struct {
	Version string
}

func (e *ErrInvalidVersion) Error() string {
	return "Version %s does not exist " + e.Version
}

// Wit struct is used to configure client
type Wit struct {
	Token   string
	URL     string
	Version string
	Verbose string
}

// New function creates default instance of Wit struct
func New() (*Wit, error) {
	DefaultWit := &Wit{
		defaults("WIT_API_TOKEN", ""),
		defaults("WIT_API_URL", "https://api.wit.ai"),
		defaults("WIT_API_VERSION", "20180527"),
		defaults("WIT_API_VERBOSE", "false"),
	}
	err := DefaultWit.SanityCheck()
	return DefaultWit, err
}

// Message function creates new message struct
func (w *Wit) Message(msg string) *Message {
	urlMessage := &url.URL{Path: msg}
	params := "/message?verbose=" + w.Verbose + "&q=" + urlMessage.String()
	return &Message{w, params}
}

// SanityCheck function checks if Wit struct attributes are valid
func (w *Wit) SanityCheck() error {
	fn := [](func() error){w.validateToken, w.validateVersion, w.validateVerbose}
	for _, f := range fn {
		err := f()
		if err != nil {
			return err
		}
	}
	return nil
}

// CreateRequest function creates http.Client and populates headers with default values
func (w *Wit) CreateRequest(method, params string) (*http.Client, *http.Request, error) {
	err := w.SanityCheck()
	if err != nil {
		return nil, nil, err
	}
	client := &http.Client{}
	req, err := http.NewRequest(method, w.URL+params, nil)
	req.Header.Add("Authorization", "Bearer "+w.Token)
	req.Header.Add("Accept", "application/vnd.wit."+w.Version+"+json")
	if err != nil {
		return nil, nil, err
	}
	return client, req, nil
}

// DoRequest function makes actual http call and return response body and return code
func (w *Wit) DoRequest(req *http.Request, client *http.Client) ([]byte, int, error) {
	resp, err := client.Do(req)
	if err != nil {
		return nil, 0, err
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, err
	}
	return body, resp.StatusCode, err
}

// MakeRequest function is wrapper around DoRequest and CreateRequest functions
func (w *Wit) MakeRequest(method, params string) ([]byte, int, error) {
	client, req, err := w.CreateRequest(method, params)
	if err != nil {
		return nil, 0, err
	}
	body, code, err := w.DoRequest(req, client)
	if err != nil {
		return nil, 0, err
	}
	return body, code, nil
}

func (w *Wit) validateToken() error {
	if w.Token == "" {
		return ErrInvalidToken
	}
	return nil
}

func (w *Wit) validateVersion() error {
	t, err := time.Parse("20060102", w.Version)
	if err != nil {
		return err
	}
	if t.String() > time.Now().String() {
		return &ErrInvalidVersion{Version: w.Version}
	}
	return nil
}

func (w *Wit) validateVerbose() error {
	if w.Verbose != "false" && w.Verbose != "true" {
		return ErrInvalidVerbose
	}
	return nil
}

func defaults(parameter, fallback string) string {
	value := os.Getenv(parameter)
	if value == "" {
		value = fallback
	}
	return value
}

// TODO Also create function to populate optional arguments.
